enterprise/server/remoteauth/remoteauth.go

```diff
@@ -0,0 +1,189 @@
+	// beyond this size, old entries will be evicted. Increasing the size of
+	// the jwt cache will result in fewer evictions (if the cache fills up) at
+	// the cost of more memory use.
+	jwtCacheSize = 10_000
+)
+
+var (
+	authHeaders = []string{authutil.APIKeyHeader}
+
+	remoteAuthTarget = flag.String("auth.remote_auth_target", "", "The gRPC target of the remote authentication API.")
+)
+
+func NewRemoteAuthenticator() (*RemoteAuthenticator, error) {
+	conn, err := grpc_client.DialSimple(*remoteAuthTarget)
+	if err != nil {
+		return nil, err
+	}
+	return newRemoteAuthenticator(conn)
+}
+
+func newRemoteAuthenticator(conn grpc.ClientConnInterface) (*RemoteAuthenticator, error) {
+	config := &lru.Config[string]{
+		MaxSize: jwtCacheSize,
+		SizeFn:  func(v string) int64 { return 1 },
+	}
+	cache, err := lru.NewLRU(config)
+	if err != nil {
+		return nil, err
+	}
+	return &RemoteAuthenticator{
+		authClient: authpb.NewAuthServiceClient(conn),
+		cache:      cache,
+	}, nil
+}
+
+type RemoteAuthenticator struct {
+	authClient authpb.AuthServiceClient
+	cache      interfaces.LRU[string]
+}
+
+// Admin stuff unsupported in remote authenticator.
+func (a *RemoteAuthenticator) AdminGroupID() string {
+	return ""
+}
+
+// TODO(iain): control via flag if needed.
+func (a *RemoteAuthenticator) AnonymousUsageEnabled(ctx context.Context) bool {
+	return false
+}
+
+// Unsupported in remote authenticator.
+func (a *RemoteAuthenticator) PublicIssuers() []string {
+	return []string{}
+}
+
+func (a *RemoteAuthenticator) Login(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP login unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) Logout(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP logout unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) Auth(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP auth unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) AuthenticatedHTTPContext(w http.ResponseWriter, r *http.Request) context.Context {
+	return r.Context()
+}
+
+func (a *RemoteAuthenticator) SSOEnabled() bool {
+	return false
+}
+
+func (a *RemoteAuthenticator) AuthenticatedGRPCContext(ctx context.Context) context.Context {
+	// Pass JWTs through as is.
+	// TODO(iain): only use valid JWTs once running beyond GCP.
+	if jwt := getJWT(ctx); jwt != "" {
+		return context.WithValue(ctx, authutil.ContextTokenStringKey, jwt)
+	}
+
+	key := getAPIKey(ctx)
+	if key == "" {
+		return ctx
+	}
+	jwt, found := a.cache.Get(key)
+	if found {
+		return context.WithValue(ctx, authutil.ContextTokenStringKey, jwt)
+	}
+	jwt, err := a.authenticate(ctx)
+	if err != nil {
+		log.Debugf("Error remotely authenticating: %s", err)
+		return ctx
+	}
+	a.cache.Add(key, jwt)
+	return context.WithValue(ctx, authutil.ContextTokenStringKey, jwt)
```

Comment:
there's `authutil.AuthContextWithError()` which might be the "correct" way to return an error from this func - e.g. oidc.go will call `AuthContextFromClaims` which use that func when there's an error

Reply:
oh nice, switched!